ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/UnifiedProtocolSchedule.java

```diff
@@ -0,0 +1,218 @@
+import java.util.TreeSet;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import com.google.common.annotations.VisibleForTesting;
+
+public class UnifiedProtocolSchedule implements ProtocolSchedule {
+
+  @VisibleForTesting
+  protected NavigableSet<ScheduledProtocolSpec> protocolSpecs =
+      new TreeSet<>(Comparator.comparing(ScheduledProtocolSpec::milestone).reversed());
+
+  private final Optional<BigInteger> chainId;
+
+  public UnifiedProtocolSchedule(final Optional<BigInteger> chainId) {
+    this.chainId = chainId;
+  }
+
+  @VisibleForTesting
+  protected UnifiedProtocolSchedule(final UnifiedProtocolSchedule protocolSchedule) {
+    this.chainId = protocolSchedule.chainId;
+    this.protocolSpecs = protocolSchedule.protocolSpecs;
+  }
+
+  @Override
+  public ProtocolSpec getByBlockHeader(final ProcessableBlockHeader blockHeader) {
+    //    checkArgument(number >= 0, "number must be non-negative");
+    checkArgument(
+        !protocolSpecs.isEmpty(), "At least 1 milestone must be provided to the protocol schedule");
+    checkArgument(
+        protocolSpecs.last().milestone() == 0, "There must be a milestone starting from block 0");
+
+    final Map<Boolean, List<ScheduledProtocolSpec>> partitionedSpecs =
+        protocolSpecs.stream().collect(partitioningBy(ScheduledProtocolSpec::isTimestampMilestone));
+    final List<ScheduledProtocolSpec> blockNumberSpecs = partitionedSpecs.get(false);
+    final List<ScheduledProtocolSpec> timestampSpecs = partitionedSpecs.get(true);
+
+    assert blockNumberSpecs.stream()
+            .mapToLong(ScheduledProtocolSpec::milestone)
+            .max()
+            .orElse(Long.MIN_VALUE)
+        < timestampSpecs.stream()
+            .mapToLong(ScheduledProtocolSpec::milestone)
+            .min()
+            .orElse(Long.MAX_VALUE);
+
+    final ScheduledProtocolSpec byBlockNumber =
+        getByMilestone(blockNumberSpecs, blockHeader.getNumber());
+    final ScheduledProtocolSpec byTimestamp =
+        getByMilestone(timestampSpecs, blockHeader.getTimestamp());
+    return Stream.of(byBlockNumber, byTimestamp)
+        .filter(Objects::nonNull)
+        .max(Comparator.comparing(ScheduledProtocolSpec::milestone))
+        .map(ScheduledProtocolSpec::spec)
+        .orElse(null);
+  }
+
+  private ScheduledProtocolSpec getByMilestone(
+      final List<ScheduledProtocolSpec> specsAscending, final long number) {
+
+    NavigableSet<ScheduledProtocolSpec> navigableSpecs =
+        new TreeSet<>(Comparator.comparingLong(ScheduledProtocolSpec::milestone).reversed());
+    navigableSpecs.addAll(specsAscending);
+
+    // protocolSpecs is sorted in descending block order, so the first one we find that's lower than
+    // the requested level will be the most appropriate spec
+    for (final ScheduledProtocolSpec s : navigableSpecs) {
+      if (number >= s.milestone()) {
+        return s;
+      }
+    }
+    return null;
+  }
```

Comment:
This is the most important change and represents a change to the algorithm for retrieving the protocol schedule: it uses the original MutableScheduleProtocol.getByBlockNumber algorithm (running through the schedule in reverse order), however it does it again for timestamp and then using the max milestone.

The reason for this is the the original algorithm needs to land on the latest schedule, if for example we only have blockNumber milestones and current block is after the latest milestone.

This algorithm is a first draft, it's not as efficient as it could be and I plan to continue working on it.